File: 01-09-2024_Max_Sum_Path_in_Two_Arrays.py
class Solution:
    def maxPathSum(self, ar1, ar2):
        m, n = len(ar1), len(ar2)
        i, j = 0, 0
    
        result, sum1, sum2 = 0, 0, 0
        
        while i < m and j < n:
            if ar1[i] < ar2[j]:
                sum1 += ar1[i]
                i += 1
            elif ar1[i] > ar2[j]:
                sum2 += ar2[j]
                j += 1
            else:
                result += max(sum1, sum2) + ar1[i]
                sum1, sum2 = 0, 0
                i += 1
                j += 1
        
        while i < m:
            sum1 += ar1[i]
            i += 1
        
        while j < n:
            sum2 += ar2[j]
            j += 1
        
        result += max(sum1, sum2)
        
        return result

File: 01-09-2024_Max_Sum_Path_in_Two_Arrays.java
class Solution {
    public int maxPathSum(List<Integer> ar1, List<Integer> ar2) {
        int m = ar1.size(), n = ar2.size();
        int i = 0, j = 0;

        int result = 0, sum1 = 0, sum2 = 0;
    
        while (i < m && j < n) {
            if (ar1.get(i) < ar2.get(j))
                sum1 += ar1.get(i++);
    
            else if (ar1.get(i) > ar2.get(j))
                sum2 += ar2.get(j++);
    
            else {
                result += Math.max(sum1, sum2) + ar1.get(i);
    
                sum1 = 0;
                sum2 = 0;
    
                i++;
                j++;
            }
        }
    
        while (i < m)
            sum1 += ar1.get(i++);
    
        while (j < n)
            sum2 += ar2.get(j++);
    
        result += Math.max(sum1, sum2);
    
        return result;
    }
    
}
